### ISSUE TITLE
Rendering Outside Perspective with a Vector

### ISSUE THREAD

**Issue by Jorkdekok1** [Jorkdekok1/CAR-system-Neurophysiology#10]
The data I saved during the novel object recognition task was already processed closed-loop a bit. Namely, the coordinates were in cm, the x axis was flipped and and there was a phase transition to align the coordinate system of the incoming data of SLEAP and the location of the virtual object on the screen. Also, I do arctan2(z,x) instead of arctan2(x,z) this means I calculate the angle relative to the Z axis instead of X, I am not sure if this matters. The following is script contains all these processing steps: 

![Image](https://github.com/Jorkdekok1/CAR-system-Neurophysiology/assets/126770475/72cab511-cb51-429a-9fa2-a640542edaf1)

Now during the reconstruction of the view, I want to compare it with the actual videos. I made a script that seems to work, however I do not understand it completely. Instead of flipping the x axis, I flipped the Z axis. The phase transition should be fixed wardAffine, however I noticed that the -50degrees is necessary to make the vector seem aligned with the video. 

![Image](https://github.com/Jorkdekok1/CAR-system-Neurophysiology/assets/126770475/c1f1c4b4-4cbc-42bb-b6c8-f625162b7dfe)

https://github.com/Jorkdekok1/CAR-system-Neurophysiology/assets/126770475/4d095af2-5c25-4ef1-aedc-ca597d8d6260

I have tried to align the vector without using the warpAffine, it would make sense if I would do +68 degrees to make the video and vector align. However this is not the case + 22 degrees does seems to align well with the video. 

https://github.com/Jorkdekok1/CAR-system-Neurophysiology/assets/126770475/0d493080-ec80-4e88-ba52-5f17c016af59

**Comment by Jorkdekok1** [comment]
I have been struggling a bit with the conversions of from cartesian, to polar, to cartesian, to polar, to cartesian. In python i get the following, I process the data similar to what I do with the input of sleap, I convert to polar coordinates do - 68 degrees and flip x axis by doing - (function called processing). After that I want to undo the processing with the second function I do + 22 degrees and i flip both the x and z axis, the results is that I get the same values as the initial trajectory (function called reverse_trajectory_processing). I do not fully understand this but it does work nevertheless. 

![Image](https://github.com/Jorkdekok1/CAR-system-Neurophysiology/assets/126770475/28c4312b-129e-456d-a9e2-2b587af8e309)

Because this works I try to do the same within bonsai to get same result. The incoming trajectory data from the csv file is already processed by the first function (-68 degrees and - x). Then I apply the second function, however when I do the same thing as in python it does not work. Instead when I do +22 degrees and (only) flip the z axis it does seem work (see video). 

![Image](https://github.com/Jorkdekok1/CAR-system-Neurophysiology/assets/126770475/af071be0-0e48-44bb-a5a7-ec58ab94b468)

https://github.com/Jorkdekok1/CAR-system-Neurophysiology/assets/126770475/f4b4547a-7cd1-4fce-ab71-f43a46243cb7

I am not really sure how to proceed, as it does seem to work but I do not understand it as I cant replicate it in python.

**Comment by Jorkdekok1** [comment]
I think I am getting a little closer to understanding this. I made one node to align the CSV trajectory with the video. An open issue with this part of the pipeline is that I do not understand why this works and is different form my pipeline in Spyder (i.e. flipping only z axis; see previous comment). 

![Image](https://github.com/Jorkdekok1/CAR-system-Neurophysiology/assets/126770475/bfb028fe-a0db-4f69-8708-f7904c64da28)

I noticed that the initial processing function besides the phase shift of 68 degrees also flips the x-axis (function called processing; see Spyder script in previous comment). So, to get to the ground truth with the video we also need to perform a horizontal flip. I did this with the flip node in Bonsai. I then just let the CSV trajectory run (which has been generated by the first Bonsai figure in this thread; i.e. -68 degrees and flipped x-axis) and looked at the video and rotated it till it looked similar. A rotation of -22 degrees seems to work well, instead of rotating the video -68 degrees. So, I still have some inconsistencies in regards to the alignment, but I am getting closer to it. 

https://github.com/Jorkdekok1/CAR-system-Neurophysiology/assets/126770475/03f8efb3-4617-46f9-bf6e-82917cbd71eb

**Comment by Jorkdekok1** [comment]
progress: The code seems to work well now as I think I aligned everything. I do not understand the mathematical background fully and it seems very weird to me that the trajectory is rotate -68 and the X-axis is flipped. Whereas the video is rotated -22 and there is a horizontal flip. Nevertheless The top-view video, vector and generated video texture seems to be all aligned now. Also, I added the radial restriction to the trajectory

https://github.com/Jorkdekok1/CAR-system-Neurophysiology/assets/126770475/2673fbeb-11f6-457e-b9c8-f14eb8cb44b3

Next steps: 
Smoothing the head direction and finding the correct orientation. The following might provide an indication. Also need to work on the delay between the vector moving and the video. 

![Image](https://github.com/Jorkdekok1/CAR-system-Neurophysiology/assets/126770475/59c6674c-409a-4122-821c-76f1bedd5546)

![Image](https://github.com/Jorkdekok1/CAR-system-Neurophysiology/assets/126770475/8bb58427-cf1f-4171-8203-8f5af8475288)

**Comment by Jorkdekok1** [comment]
Latest update: In my previous troubleshoots I tried to get to reverse the initial processing (-68 degrees and -x; first image in this comment) by firstly doing a rotation and then flipping axis. This did not make much sense as it would be better to first flip and then do the rotation (i.e. the reverse of the initial processing). however, the numbers still do not add up as I cannot reverse the processing by doing -x and +68 degrees (see fig2 and vid1). To align the trajectory and video i did -z followed by -22 degrees (fig3 and vid2). Another way to verify this result is by aligning the video with the original trajectory data (which went through the initial processing pipeline; i.e. Fig1). The video with also align with this trajectory if i do -22 degrees and -Z (Fig4). 

It seems to be the case that somehow -Z followed by -22 are equivalent to -68 followed by -x, I think the problem is that the X and Z are differently defined in video frames. In regards to the incoming video data, up is defined as -Z and down Z; X is right and -X left (Fig 5). I guess this might have something to do with it, however I did try and draw every operation out and I do not get the equivalent back (Fig 6). 

Also, confusingly, In Spyder using the initial function can be refers by done -x followed by -68. This might indicate that something is happening in the original pipeline that is different. However I do not see anywhere where i might have switched around the around or something like that. 

Lastly, I think it might not be possible to use realign the orginal to fit the video as the virtual environment would also need to be processed as well. I can do a rotation that is no problem, however the would also need to -Z I would not know were to start with. With the texture or the blender model of the screen. The other way (i.e. fitting the video to the processed trajectory) I can already do and also align it with the video texture on the blender model of the screen. 

![image](https://github.com/Jorkdekok1/CAR-system-Neurophysiology/assets/126770475/861d507b-5422-47fb-bbc0-3fac399e972a)
Fig1.

![image](https://github.com/Jorkdekok1/CAR-system-Neurophysiology/assets/126770475/9ebda708-91cc-41c6-837b-cb6d6ec17a14)
Fig2.

https://github.com/Jorkdekok1/CAR-system-Neurophysiology/assets/126770475/e16c92ae-9cb9-4cb0-be95-7c16b5858969
Vid1.

![image](https://github.com/Jorkdekok1/CAR-system-Neurophysiology/assets/126770475/90e1967c-82bb-4997-b031-edc4a099ca26)
Fig 3.

https://github.com/Jorkdekok1/CAR-system-Neurophysiology/assets/126770475/c49da600-d8de-4364-b72e-bf3eb006dd7a
Vid2.

![image](https://github.com/Jorkdekok1/CAR-system-Neurophysiology/assets/126770475/e18c2301-d1c2-4cd1-99a8-3d640a76782e)
Fig4. 

![image](https://github.com/Jorkdekok1/CAR-system-Neurophysiology/assets/126770475/67e0880e-d42c-4d4f-9114-087c3433590a)
Fig5.

![image](https://github.com/Jorkdekok1/CAR-system-Neurophysiology/assets/126770475/f4c2deb2-e558-4a9c-a00e-6b41576c7804)
Fig6. 

![image](https://github.com/Jorkdekok1/CAR-system-Neurophysiology/assets/126770475/0404bf3d-6224-4c1c-b2ff-2524f04961a7)
Fig7.

**Comment by amansaleem** [comment]
If you had to start this again with a new animal, would it be possible, if you saved the correct information?

Is this a fundamental issue, or just an issue because you potentially saved the information at a non-ideal point?

**Comment by Jorkdekok1** [comment]
I think the best solution is to update the current pipeline and align the virtual environments by using warp affine and flip instead of using the python node in Bonsai. The SLEAP model would need to be retrained on the video where the warp and flip is implemented, but we need to train a new model anyway to track mice. I already adjusted the pipeline and used the car to check everything (Fig1). When I do a rotation of +68 and flip horizontally the tracking is aligned with the virtual environment (Fig2 and Vid1). I checked just to be sure and the python node I use previously used (which does -68 and -X) provide the same results when the video is not warped and flipped (looked the same as Vid1). 

![image](https://github.com/Jorkdekok1/CAR-system-Neurophysiology/assets/126770475/434df424-04ee-4bde-9d18-2bfd9bf6fa0e)
Fig1.

![image](https://github.com/Jorkdekok1/CAR-system-Neurophysiology/assets/126770475/e572f3dd-fe24-40c5-afb4-7da745b06f2c)
Fig2.

https://github.com/Jorkdekok1/CAR-system-Neurophysiology/assets/126770475/a4075e6a-1df0-490c-b4af-40174c1e177d
Vid1.

Saving the raw video and raw unwarped/flipped SLEAP data might not be the best option as the data and video would still need to be aligned with the VE. And I do seem to get different results altough some kind of symmetry in the different rotations and flips I did during my troubleshoots, so somehow something in Bonsai does the right thing but by implementing different values. So, to circumvent this problem,, I think it is best to save the video where the video is already flipped and warped then the SLEAP data and video are already aligned (as in Fig1) I think then there will be no problem in the future.
